teos/watcher.py

```diff
@@ -19,6 +21,100 @@ class AppointmentLimitReached(BasicException):
+            Used to keep track of what data belongs to what block, so data can be pruned accordingly. Also needed to
+            rebuild the cache in case of reorgs.
+        cache_size (:obj:`int`): the size of the cache in blocks.
+    """
+
+    def __init__(self, blocks_in_cache):
+        self.cache = dict()
+        self.blocks = OrderedDict()
+        self.cache_size = blocks_in_cache
+
+    def init(self, last_known_block, block_processor):
+        """
+        Sets the initial state of the locator cache.
+
+        Args:
+            last_known_block (:obj:`str`): the last known block by the ``Watcher``.
+            block_processor (:obj:`teos.block_processor.BlockProcessor`): a ``BlockProcessor`` instance.
+        """
+
+        # This is needed as a separate method from __init__ since it has to be initialized right before start watching.
+        # Not doing so implies store temporary variables in the Watcher and initialising the cache as None.
+        target_block_hash = last_known_block
+        for _ in range(self.cache_size):
+            # In some setups, like regtest, it could be the case that there are no enough previous blocks.
+            # In those cases we pull as many as we can (up to cache_size).
+            if target_block_hash:
+                target_block = block_processor.get_block(target_block_hash)
+                if not target_block:
+                    break
+            else:
+                break
+
+            locator_txid_map = {compute_locator(txid): txid for txid in target_block.get("tx")}
+            self.cache.update(locator_txid_map)
+            self.blocks[target_block_hash] = list(locator_txid_map.keys())
+            target_block_hash = target_block.get("previousblockhash")
+
+        self.blocks = OrderedDict(reversed((list(self.blocks.items()))))
+
+    def fix_cache(self, last_known_block, block_processor):
+        """
+        Fixes an existing cache after a reorg has been detected by feeding the last ``cache_size`` blocks to it.
+
+        Args:
+            last_known_block (:obj:`str`): the last known block hash after the reorg.
+            block_processor (:obj:`teos.block_processor.BlockProcessor`): a ``BlockProcessor`` instance.
+        """
+
+        tmp_cache = LocatorCache(self.cache_size)
+
+        # We assume there are no reorgs back to genesis. If so, this would raise some log warnings. And the cache will
+        # be filled with less than cache_size blocks.
+        target_block_hash = last_known_block
+        for _ in range(tmp_cache.cache_size):
+            target_block = block_processor.get_block(target_block_hash)
+            if target_block:
+                locator_txid_map = {compute_locator(txid): txid for txid in target_block.get("tx")}
+                tmp_cache.cache.update(locator_txid_map)
+                tmp_cache.blocks[target_block_hash] = list(locator_txid_map.keys())
+                target_block_hash = target_block.get("previousblockhash")
```

Comment:
There are only two reasons why we won't find a block in this case: a temporal `bitcoind` error or the block not being there.

In the former we may actually want to repeat the query, since it may return the block (i've never seen one of this though, but I assume it can happen). The later may only happen if `bitcoind` was running on pruning mode. If that's the case the code may have failed long before, since we need full mode with txindex.

I'll add some comments so the loop it's easier to follow though